components/pd_client/src/util.rs

```diff
@@ -305,24 +297,24 @@ where
         }
     }
 
-    fn post_loop(ctx: Result<Self>) -> Result<Resp> {
-        let ctx = ctx.expect("end loop with Ok(_)");
-        ctx.resp
+    fn post_loop(self) -> Result<Resp> {
+        self.resp
             .unwrap_or_else(|| Err(box_err!("response is empty")))
     }
 
     /// Returns a Future, it is resolves once a future returned by the closure
     /// is resolved successfully, otherwise it repeats `retry` times.
-    pub fn execute(self) -> PdFuture<Resp> {
-        let ctx = self;
-        Box::new(
-            loop_fn(ctx, |ctx| {
-                ctx.reconnect_if_needed()
-                    .and_then(Self::send_and_receive)
-                    .then(Self::break_or_continue)
-            })
-            .then(Self::post_loop),
-        )
+    pub fn execute(mut self) -> PdFuture<Resp> {
+        Box::pin(async move {
+            loop {
+                while self.reconnect_if_needed().await {}
```

Comment:
The behavior here is different from before. If reconnection fails forever, this will loop forever. However, in the previous code, it will return error when `reconnect_count` decreases to zero.